tachyon/zk/plonk/permutation/permutation_assembly.h

```diff
@@ -0,0 +1,151 @@
+// Copyright 2020-2022 The Electric Coin Company
+// Copyright 2022 The Halo2 developers
+// Use of this source code is governed by a MIT/Apache-2.0 style license that
+// can be found in the LICENSE-MIT.halo2 and the LICENCE-APACHE.halo2
+// file.
+
+#ifndef TACHYON_ZK_PLONK_PERMUTATION_PERMUTATION_ASSEMBLY_H_
+#define TACHYON_ZK_PLONK_PERMUTATION_PERMUTATION_ASSEMBLY_H_
+
+#include <utility>
+#include <vector>
+
+#include "tachyon/base/containers/container_util.h"
+#include "tachyon/base/openmp_util.h"
+#include "tachyon/math/polynomials/univariate/univariate_evaluation_domain.h"
+#include "tachyon/zk/plonk/permutation/cycle_store.h"
+#include "tachyon/zk/plonk/permutation/label.h"
+#include "tachyon/zk/plonk/permutation/lookup_table.h"
+#include "tachyon/zk/plonk/permutation/permutation_argument.h"
+#include "tachyon/zk/plonk/permutation/permutation_proving_key.h"
+#include "tachyon/zk/plonk/permutation/permutation_verifying_key.h"
+
+namespace tachyon::zk {
+
+// Struct that accumulates all the necessary data in order to construct the
+// permutation argument.
+template <typename Curve, size_t MaxDegree>
+class PermutationAssembly {
+ public:
+  constexpr static size_t kRows = MaxDegree + 1;
+
+  using ScalarField = typename Curve::ScalarField;
+  using Evals = math::UnivariateEvaluations<ScalarField, MaxDegree>;
+  using DensePoly = math::UnivariateDensePolynomial<ScalarField, MaxDegree>;
+
+  // It has parameters for commitment and provide commit function.
+  // TODO(dongchangYoo): substitute it to KZGParams, but not implemented yet.
+  using ParamsTy = std::vector<math::AffinePoint<Curve>>;
+
+  PermutationAssembly() = default;
+
+  // Constructor with |PermutationArgument| which has copy permutation columns.
+  explicit PermutationAssembly(const PermutationArgument& p)
+      : PermutationAssembly(p.columns()) {}
+
+  // Constructor with permutation columns.
+  explicit PermutationAssembly(const std::vector<AnyColumn>& columns)
+      : columns_(columns), cycle_store_(CycleStore(columns.size(), kRows)) {}
```

Comment:
For consistency with below,

```suggestion
      : columns_(columns), cycle_store_(CycleStore(columns_.size(), kRows)) {}
```